fsspec/asyn.py

```diff
@@ -239,21 +239,19 @@ async def _run_coros_in_chunks(
         batch_size = len(coros)
 
     assert batch_size > 0
-    results = []
-    for start in range(0, len(coros), batch_size):
-        chunk = [
-            asyncio.Task(asyncio.wait_for(c, timeout=timeout))
-            for c in coros[start : start + batch_size]
-        ]
+    semaphore = asyncio.BoundedSemaphore(batch_size)
+
+    async def _worker(coro):
+        async with semaphore:
```

Comment:
With this approach, I agree that a limited number of IO tasks are concurrently running/waiting, but is there any potential downside of simply having a very large number of tasks created and waiting on this semaphore? The previous incantation would only have the given number of top-level tasks at a time.

Reply:
Good question. I'm not aware of any downsides but can't prove there aren't any. I was also considering to use [`asyncio.as_completed`](https://docs.python.org/3/library/asyncio-task.html#asyncio.as_completed) and pass an iterable of `asyncio.Task`s, but it also materializes this iterator as a set internally.

Reply:
Let's leave this question open a little while, to see if any knowledgable person comments.

Reply:
FWIW, I've used this implementation to limit the simultaneous execution of 70,000 coroutines and it worked fine.

Reply:
Would you like to try making a queue? It might turn out to be simpler.

Reply:
https://death.andgravity.com/limit-concurrency This recent blogpost has some really nice solutions to this problem.

Reply:
Thanks for the pointers, @Skylion007! 👌 I got a queue-based solution working today except when `return_exceptions=False`. But the blog article you referenced offers a much more comprehensive analysis and different solution.

Reply:
I've pushed an implementation based on `asyncio.wait` inspired by https://death.andgravity.com/limit-concurrency#asyncio-wait. I think this looks good and tests are passing.

Reply:
I actually think the linked solution may be more efficient. Does this solution run into the issue that the next batch does not start until the current batch finishes? That means with large batch size, future workers will be stalled waiting on the slowest worker?

Or am I misunderstanding the current batch async method.

Reply:
> Does this solution run into the issue that the next batch does not start until the current batch finishes? That means with large batch size, future workers will be stalled waiting on the slowest worker?

Correct, the slowest coroutine in a batch determines the runtime of the whole batch and the next batch won't start before all coroutines in the current batch have finished.

Reply:
I think some of the solutions I linked would have alleviated that issue so it could have eagerly continued the next batch.

Reply:
The solution that @martindurant merged is essentially https://death.andgravity.com/limit-concurrency#asyncio-wait, which you linked.